File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    void recordOneCircle(vector<vector<int>>& matrix, int step, vector<int>& ans) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        ans.push_back(matrix[step][step]);
        
        for (int j = step + 1; j <= n - step - 1; j++) {
            int row = step;
            ans.push_back(matrix[row][j]);
        }
        
        for (int i = step + 1; i <= m - step - 1; i++) {
            int col = n - step - 1;
            ans.push_back(matrix[i][col]);
        }
        
        if (step < m - step - 1) {
            for (int j = n - step - 2; j >= step; j--) {
                int row = m - step - 1;
                ans.push_back(matrix[row][j]);
            }
        }

        if (step < n - step - 1) {
            for (int i = m - step - 2; i >= step + 1; i--) {
                int col = step;
                ans.push_back(matrix[i][col]);
            }
        }
    }

    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int min_side = min(m, n);

        vector<int> ans;
        int step = 0;

        while (min_side - step*2 > 0) {
            recordOneCircle(matrix, step, ans);
            step++;
        }
        
        return ans;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        cells = m * n
        
        left_most = 0
        right_most = n - 1
        top_most = 1
        bottom_most = m - 1

        ans = [matrix[0][0]]
        ci = 0
        cj = 0
        count = 1

        while True:
            # go right
            while cj < right_most:
                cj += 1
                ans.append(matrix[ci][cj])
                count += 1
            right_most -= 1
            if count == cells:
                break

            # go down
            while ci < bottom_most:
                ci += 1
                ans.append(matrix[ci][cj])
                count += 1
            bottom_most -= 1
            if count == cells:
                break

            # go left
            while cj > left_most:
                cj -= 1
                ans.append(matrix[ci][cj])
                count += 1
            left_most += 1
            if count == cells:
                break
            
            # go up
            while ci > top_most:
                ci -= 1
                ans.append(matrix[ci][cj])
                count += 1
            top_most += 1
            if count == cells:
                break

        return ans